### PR TITLE
SRIOV (w/wo Offload) new network configurations

### PR DESCRIPTION
**Pull request by zxa13**
Included the OVS-DPDK network configurations for the following cases:

- SRIOV 6 port
This will enable 2 port SRIOV w/o Offload when a separate provisioning interface is not used
- SRIOV 8 port
This will enable 4 port SRIOV w/o Offload when a separate provisioning interface is not used
- SRIOV 6 port w/ Offload
This will enable 2 port SRIOV w/ Offload when a separate provisioning interface is not used
- SRIOV 7 port w/ Offload
This will enable 2 port SRIOV w/ Offload when a separate provisioning interface is used
- SRIOV 8 port w/ Offload
This will enable 4 port SRIOV w/ Offload when a separate provisioning interface is not used
- SRIOV 9 port w/ Offload
This will enable 4 port SRIOV w/ Offload when a separate provisioning interface is used

### CODE REVIEW

**Comment by zxa13** [comment]
@cdearborn @heydog15 Kindly review and merge. Thanks

File: src/deploy/osp_deployer/profiles/csp.json [hunk 1]
```diff
@@ -231,7 +239,7 @@
                   "valid_values": [
                   "true",
                   "false"
-               ],
+               ]
```
Comment (cdearborn) [inline src/deploy/osp_deployer/profiles/csp.json]: The change on the above line introduces a syntax error to this file.  Need to undo this.

File: src/deploy/osp_deployer/profiles/csp.json [hunk 2]
```diff
@@ -249,8 +257,13 @@
                ]
             },
          "use_ipmi_driver":"",
-         "use_in_band_introspection":"",
```
Comment (cdearborn) [inline src/deploy/osp_deployer/profiles/csp.json]: It doesn't look to me like the changes in this block are related to this patch.  We should undo these.

File: src/deploy/osp_deployer/director.py [hunk 3]
```diff
@@ -1289,31 +1289,24 @@ def setup_sriov_nic_configuration(self):
...
         sriov_map_setting = []
         sriov_pci_passthrough = []
-        physical_network = "physint"
-        physical_network1 = "tenant1"
-        physical_network2 = "tenant2"
-        check = 1
-        for interface in sriov_interfaces:
+        physical_network = ["physint", "physint1"]
+        check = 0
+        for interface in sriov_interfaces:   # Do it for only first two interfaces, if smart nic is enabled. otherwise
```
Comment (jtwill98) [inline src/deploy/osp_deployer/director.py]: 53 additional pep8 compliance violations have been introduced with this PR to director.py.  I recommend these be addressed.  The following were present prior to the PR and is provided here as an example of how to run pep8 and to view the errors.  

pep8 director.py #where Ennn=Error and Wnnn=Warnings. 

director.py:167:16: E111 indentation is not a multiple of four
director.py:167:80: E501 line too long (82 > 79 characters)
director.py:169:16: E111 indentation is not a multiple of four
director.py:171:16: E111 indentation is not a multiple of four
director.py:171:80: E501 line too long (81 > 79 characters)
director.py:173:16: E111 indentation is not a multiple of four
director.py:175:1: W293 blank line contains whitespace
director.py:177:5: E303 too many blank lines (3)
director.py:442:46: W605 invalid escape sequence '\$'
director.py:967:38: W605 invalid escape sequence '\ '
director.py:985:30: W605 invalid escape sequence '\ '

File: src/deploy/osp_deployer/settings_sanity.py [hunk 4]
```diff
@@ -585,3 +585,11 @@ def verify_sriov_dependencies(self):
             raise AssertionError("SR-IOV and DVR can not be " +
                                  "enabled together.Please disable" +
                                  "one feature and verify the settings.")
+
+    def verify_hw_offload_dependencies(self):
+        logger.debug("verifying Smart NIC is enabled when SR-IOV is enabled")
+        if (self.settings.enable_sriov is False and
+                self.settings.enable_smart_nic is True):
+            raise AssertionError("SR-IOV hardware offload cant be " +
```
Comment (cdearborn) [inline src/deploy/osp_deployer/settings_sanity.py]: nit: in the line above, change "cant" to "can't"

File: src/deploy/osp_deployer/settings_sanity.py [hunk 5]
```diff
@@ -585,3 +585,11 @@ def verify_sriov_dependencies(self):
...
+                                 "enabled as SRIOV is not being enabled. " +
```
Comment (cdearborn) [inline src/deploy/osp_deployer/settings_sanity.py]: nit: in the line above, remove "being"

File: src/deploy/osp_deployer/settings_sanity.py [hunk 6]
```diff
@@ -585,3 +585,11 @@ def verify_sriov_dependencies(self):
             raise AssertionError("SR-IOV and DVR can not be " +
                                  "enabled together.Please disable" +
                                  "one feature and verify the settings.")
+
+    def verify_hw_offload_dependencies(self):
+        logger.debug("verifying Smart NIC is enabled when SR-IOV is enabled")
```
Comment (cdearborn) [inline src/deploy/osp_deployer/settings_sanity.py]: nit: change the above log message to:
Verifying SR-IOV is enabled when Smart NIC is enabled

File: src/deploy/osp_deployer/director.py [hunk 7]
```diff
@@ -1408,6 +1427,7 @@ def delete_overcloud(self):
         for node in ls_nodes:
             node_state = node.split("|")[5]
             node_id = node.split("|")[1]
+            nod_id = node.split("|")[1]
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: Looks like the variable added in the above line is unused.

File: src/deploy/osp_deployer/director.py [hunk 8]
```diff
@@ -1256,13 +1256,18 @@ def setup_nic_configuration(self):
...
         for cmd in cmds:
             self.run(cmd)
 
     def setup_sriov_nic_configuration(self):
         logger.debug("setting SR-IOV environment")
-        # Get seetings from .ini file
+        # Get settings from .ini file
         ini_nics_settings = self.settings.get_curated_nics_settings()
 
         cmds = []
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: The code below really should call get_sriov_interfaces() instead of duplicating that code here

File: src/deploy/osp_deployer/director.py (hunk 7 above)
Comment (jtwill98) [inline src/deploy/osp_deployer/director.py]: Why is this needed?  If needed, I recommend changing the name as the name is too similar to node_id.

File: src/deploy/osp_deployer/director.py [hunk 9]
```diff
@@ -1284,16 +1289,27 @@ def setup_sriov_nic_configuration(self):
...
         sriov_pci_passthrough = []
-        physical_network = "physint"
-        for interface in sriov_interfaces:
+        physical_network = ["physint", "physint1"]
+        check = 0
+        for interface in sriov_interfaces:   # Do it for only first two interfaces, if smart nic is enabled. otherwise
             devname = interface
-            mapping = physical_network + ':' + interface
-            sriov_map_setting.append(mapping)
+            if self.settings.enable_smart_nic == True and self.get_sriov_interfaces() == 2: # sriov_interfaces contain only two interfaces
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: So this looks line smart NICs will be enabled only if there are 2 SRIOV interfaces.  What if there are 4?  What is the behavior that we want then?  With this code, it looks like no smart NICs will be enabled.

File: src/deploy/osp_deployer/director.py [hunk 10]
```diff
@@ -1256,13 +1256,18 @@ def setup_nic_configuration(self):
             cmds.append('sed -i -r \'s/(^\s*' + setting_name +    # noqa: W605
                         ':\s*).*/\\1' + setting_value + '/\' ' + remote_file)
 
+        # if smart NICs are used then use smart configuration for contoller and compute nodes
+        if self.settings.enable_smart_nic == True and self.get_sriov_interfaces() == 2:
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: We really shouldn't be modifying the files that the nic_environment.yaml points to on the fly.  Instead, we should just add a new sriov_6_port_smart NIC configuration

**Comment by zxa13** [comment]
Okay @cdearborn and @jtwill98 thank you for the detailed review. I am working on it and update the code accordingly ASAP.

File: src/deploy/osp_deployer/profiles/xsp.json [hunk 11]
```diff
@@ -38,6 +38,13 @@
                    "false"
                 ]
              },
+
+        "smart_nic": {
+                "valid_values": [
+                   "true",
```
Comment (cdearborn) [inline src/deploy/osp_deployer/profiles/xsp.json]: Since we list only false as a valid value for sriov_enable, we should probably list only false as a valid value for smart_nic in the xsp profile

File: src/deploy/osp_deployer/director.py [hunk 12]
```diff
@@ -353,6 +353,9 @@ def node_introspection(self):
                                  "error: {}, stdout: {}".format(exit_status,
                                                                 stderr,
                                                                 stdout))
+        # Turning LLDP off before introspection
+        lldp_off_cmd = "sudo sed -i 's/ipa-collect-lldp=1/ipa-collect-lldp=0/g' /httpboot/inspector.ipxe"
+        self.run(lldp_off_cmd)
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: Can you tell me why we need to turn off LLDP prior to running introspection?

File: src/deploy/osp_deployer/director.py [hunk 13]
```diff
@@ -890,14 +892,14 @@ def setup_unity_cinder(self, dell_unity_cinder_yaml):
             self.settings.unity_storage_pool_names + '|" ' +
             dell_unity_cinder_yaml,
         ]
-
+
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: Please remove the added whitespace on the line above.

File: src/deploy/osp_deployer/director.py [hunk 14]
```diff
@@ -932,35 +934,34 @@ def setup_unity_manila(self, unity_manila_yaml):
...
-            self.settings.manila_unity_ethernet_ports + '|" ' +
-            unity_manila_yaml,
-            'sed -i "s|<manila_unity_ssl_cert_verify>|' +
-            self.settings.manila_unity_ssl_cert_verify + '|" ' +
-            unity_manila_yaml,
-            'sed -i "s|<manila_unity_ssl_cert_path>|' +
-            self.settings.manila_unity_ssl_cert_path + '|" ' +
-            unity_manila_yaml,
-        ]
+        cmds = ['sed -i "s|<manila_unity_driver_handles_share_servers>|' +
```
Comment (cdearborn) [inline src/deploy/osp_deployer/director.py]: Changes like this make it difficult and time consuming to review a patch.  It doesn't look like there were any changes made other than code reformatting, but it's difficult to tell.  Please create a separate patch for changes that only involve code formatting changes.

File: src/deploy/osp_deployer/director.py (hunk 12 above)
Comment (WajeehaHamid) [reply src/deploy/osp_deployer/director.py]: For OS10 and Mellanox NICs, we were facing LLDP related errors in in-band introspection. Disabling it resolved these issue. We are still investigating this issue that is why we added this temporary workaround.

File: src/deploy/osp_deployer/director.py [hunk 15]
```diff
@@ -353,6 +353,9 @@ def node_introspection(self):
                                  "error: {}, stdout: {}".format(exit_status,
                                                                 stderr,
                                                                 stdout))
+        # Turning LLDP off before introspection
+        lldp_off_cmd = "sudo sed -i 's/ipa-collect-lldp=1/ipa-collect-lldp=0/g' /httpboot/inspector.ipxe"
```
Comment (jtwill98) [inline src/deploy/osp_deployer/director.py]: Does not pass the pep8 compliance tests.

File: src/deploy/osp_deployer/profiles/xsp.json (hunk 11 above)
Comment (zxa13) [reply src/deploy/osp_deployer/profiles/xsp.json]: Okay will change it.

File: src/deploy/osp_deployer/director.py (hunk 14 above)
Comment (zxa13) [reply src/deploy/osp_deployer/director.py]: Yes you are right will take care of this in the future
